// @auth0/auth0-angular/esm2015/lib/abstract-navigator.js
import { Injectable, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { Location } from '@angular/common';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class AbstractNavigator {
    constructor(location, injector) {
        this.location = location;
        try {
            this.router = injector.get(Router);
        }
        catch (_a) { }
    }
    /**
     * Navigates to the specified url. The router will be used if one is available, otherwise it falls back
     * to `window.history.replaceState`.
     * @param url The url to navigate to
     */
    navigateByUrl(url) {
        if (this.router) {
            setTimeout(() => {
                this.router.navigateByUrl(url);
            }, 0);
            return;
        }
        this.location.replaceState(url);
    }
}
AbstractNavigator.ɵprov = i0.ɵɵdefineInjectable({ factory: function AbstractNavigator_Factory() { return new AbstractNavigator(i0.ɵɵinject(i1.Location), i0.ɵɵinject(i0.INJECTOR)); }, token: AbstractNavigator, providedIn: "root" });
AbstractNavigator.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
AbstractNavigator.ctorParameters = () => [
    { type: Location },
    { type: Injector }
];
